File: monolith-spigot/src/main/java/gg/warcraft/monolith/spigot/menu/SpigotMenuMapper.java
package gg.warcraft.monolith.spigot.menu;

import com.google.inject.Inject;
import gg.warcraft.monolith.api.menu.Button;
import gg.warcraft.monolith.api.menu.Menu;
import gg.warcraft.monolith.api.world.item.Item;
import gg.warcraft.monolith.api.world.item.ItemService;
import gg.warcraft.monolith.app.menu.SkullButton;
import gg.warcraft.monolith.spigot.world.item.SpigotItemMapper;
import org.bukkit.Material;
import org.bukkit.Server;
import org.bukkit.SkullType;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.SkullMeta;

import java.util.UUID;

public class SpigotMenuMapper {
    private final Server server;
    private final ItemService itemService;
    private final SpigotItemMapper itemMapper;

    @Inject
    public SpigotMenuMapper(Server server, ItemService itemService, SpigotItemMapper itemMapper) {
        this.server = server;
        this.itemService = itemService;
        this.itemMapper = itemMapper;
    }

    public Inventory map(Menu menu, UUID viewerId) {
        Player player = server.getPlayer(viewerId);
        MonolithMenuHolder menuHolder = new MonolithMenuHolder(player);
        Inventory inventory = server.createInventory(menuHolder, menu.getSize().getNumberOfSlots(), menu.getTitle());
        menu.getButtons().forEach((slot, button) -> {
            ItemStack item = map(button);
            inventory.setItem(slot, item);
        });
        return inventory;
    }

    public ItemStack mapSkull(SkullButton button) {
        ItemStack itemStack = new ItemStack(Material.PLAYER_HEAD);
        SkullMeta itemMeta = (SkullMeta) itemStack.getItemMeta();
        itemMeta.setDisplayName(button.getTitle());
        itemMeta.setLore(button.getTooltip());
        itemMeta.setOwner(button.getPlayerName());
        itemMeta.addItemFlags(ItemFlag.values());
        itemStack.setItemMeta(itemMeta);
        return itemStack;
    }

    public ItemStack map(Button button) {
        if (button instanceof SkullButton) {
            return mapSkull((SkullButton) button);
        }

        Item item = itemService.create(button.getIcon())
                .withName(button.getTitle())
                .withTooltip(button.getTooltip().toArray(new String[0]))
                .withHideAttributes(true);

        return itemMapper.map(item);
    }
}

File: monolith-spigot/src/main/scala/gg/warcraft/monolith/spigot/menu/SpigotMenuMapper.scala
/*
 * MIT License
 *
 * Copyright (c) 2020 WarCraft <https://github.com/WarCraft>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package gg.warcraft.monolith.spigot.menu

import java.util.UUID

import gg.warcraft.monolith.api.menu.Menu
import gg.warcraft.monolith.spigot.player.SpigotPlayer
import org.bukkit.inventory.{Inventory, InventoryHolder}
import org.bukkit.Server

private class SpigotMenuHolder(player: SpigotPlayer) extends InventoryHolder {
  override def getInventory: Inventory = player.getInventory
}

class SpigotMenuMapper(implicit server: Server, buttonMapper: SpigotButtonMapper) {
  def map(menu: Menu, playerId: UUID): Inventory = {
    val player = server getPlayer playerId
    val holder = new SpigotMenuHolder(player)
    val inventory = server createInventory (holder, menu.size.slots, menu.title)
    menu.buttons foreach {
      case (slot, button) =>
        val item = buttonMapper map button
        inventory setItem (slot, item)
    }
    inventory
  }
}
